modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,343 @@
+// subTree.
+const merkleRootsCacheHeight = 7
+
+type (
+	// merkleRoots is a helper struct that makes it easier to add/insert/remove
+	// merkleRoots within a SafeContract.
+	merkleRoots struct {
+		// cachedSubTrees are cached trees that can be used to more efficiently
+		// compute the merkle root of a contract. The cached trees are not
+		// persisted and are computed after startup.
+		cachedSubTrees []*cachedSubTree
+		// uncachedRoots contains the sector roots that are not part of a
+		// cached subTree. The uncachedRoots slice should never get longer than
+		// 2^merkleRootsCacheHeight since that would simply result in a new
+		// cached subTree in cachedSubTrees.
+		uncachedRoots []crypto.Hash
+
+		// file is the file of the safe contract that contains the root.
+		file *os.File
+		// numMerkleRoots is the number of merkle roots in file.
+		numMerkleRoots int
+	}
+
+	cachedSubTree struct {
+		height int
+		sum    crypto.Hash
+	}
+)
+
+// loadExistingMerkleRoots reads creates a merkleRoots object from existing
+// merkle roots.
+func loadExistingMerkleRoots(file *os.File) (mr *merkleRoots, err error) {
+	mr = &merkleRoots{
+		file: file,
+	}
+	// Get the number of roots stored in the file.
+	mr.numMerkleRoots, err = mr.lenFromFile()
+	if err != nil {
+		return
+	}
+	// Seek to the first root's offset.
+	if _, err = file.Seek(rootIndexToOffset(0), io.SeekStart); err != nil {
+		return
+	}
+	// Read the roots from the file without reading all of them at once.
+	for i := 0; i < mr.numMerkleRoots; i++ {
+		var root crypto.Hash
+		if _, err = io.ReadFull(file, root[:]); err == io.EOF {
+			break
+		} else if err != nil {
+			return
+		}
+
+		// Append the root to the unachedRoots
+		mr.uncachedRoots = append(mr.uncachedRoots, root)
+
+		// If the uncachedRoots grew too large we add them to the cache.
+		if len(mr.uncachedRoots) == (1 << merkleRootsCacheHeight) {
+			st := newCachedSubTree(mr.uncachedRoots)
+			mr.cachedSubTrees = append(mr.cachedSubTrees, st)
+			mr.uncachedRoots = mr.uncachedRoots[:0]
+		}
+	}
+	return mr, nil
+}
+
+// newCachedSubTree creates a cachedSubTree from exactly
+// 2^merkleRootsCacheHeight roots.
+func newCachedSubTree(roots []crypto.Hash) *cachedSubTree {
+	// Sanity check the input length.
+	if len(roots) != (1 << merkleRootsCacheHeight) {
+		build.Critical("can't create a cached subTree from the provided number of roots")
+	}
+	return &cachedSubTree{
+		height: int(merkleRootsCacheHeight + sectorHeight),
+		sum:    cachedMerkleRoot(roots),
+	}
+}
+
+// newMerkleRoots creates a new merkleRoots object. This doesn't load existing
+// roots from file and will assume that the file doesn't contain any roots.
+// Don't use this on a file that contains roots.
+func newMerkleRoots(file *os.File) *merkleRoots {
+	return &merkleRoots{
+		file: file,
+	}
+}
+
+// rootIndexToOffset calculates the offset of the merkle root at index i from
+// the beginning of the contract file.
+func rootIndexToOffset(i int) int64 {
```

Comment:
this function name took me a few reads to understand what it was saying. I understand it now, but maybe there's a different name which isn't as confusing on the first read.